Make grpc compile on win64bit

jtattermusch opened the pull request:
"long" type is 32bit on Win x64, which brings some trouble.

Please note the new public function gpr_int64toa.

jtattermusch commented:
Travis failure unrelated: #4295

jtattermusch commented:
Interop test failure seems unrelated, @murgatroid99  is working on fixing the problem.

jtattermusch commented:
friendly ping.

nicolasnoble commented on include/grpc++/impl/rpc_service_method.h:
```diff
@@ -251,7 +252,11 @@ class RpcService {
-  int GetMethodCount() const { return methods_.size(); }
+  int GetMethodCount() const {
+    // On win x64, int is only 32bit
+    GPR_ASSERT(methods_.size() <= INT_MAX);
```
*Cough* at least this isn't 16 bits *COUGH* *COUGH*

nicolasnoble commented on src/core/iomgr/tcp_windows.c:
```diff
@@ -281,19 +281,20 @@ static void win_write(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep,
   for (i = 0; i < tcp->write_slices->count; i++) {
-    buffers[i].len = GPR_SLICE_LENGTH(tcp->write_slices->slices[i]);
+    GPR_ASSERT(GPR_SLICE_LENGTH(tcp->write_slices->slices[i]) <= UINT32_MAX);
+    buffers[i].len = (ULONG) GPR_SLICE_LENGTH(tcp->write_slices->slices[i]);
```
size_t len = GPR_SLICE_LENGTH(tcp->write_slices->slices[i]);
GPR_ASSERT(len <= ULONG_MAX);
buffers[i].len = len;

jtattermusch replied:
Done.

jtattermusch commented:
h2_proxy_simple_request_test
h2_proxy_simple_request_nosec_test  seem to be consistently failing. Could that be because of this PR?

jtattermusch commented:
Looks like h2_proxy_simple_request_test and h2_proxy_simple_request_nosec_test is also failing on master here: https://grpc-testing.appspot.com/job/gRPC_master/13556/#showFailuresLink

jtattermusch commented:
I think I've addressed all the comments. PTAL.

ctiller commented on src/core/iomgr/tcp_windows.c:
```diff
@@ -197,7 +197,7 @@ static void win_read(grpc_exec_ctx *exec_ctx, grpc_endpoint *ep,
   tcp->read_slice = gpr_slice_malloc(8192);
 
-  buffer.len = GPR_SLICE_LENGTH(tcp->read_slice);
+  buffer.len = (ULONG)GPR_SLICE_LENGTH(tcp->read_slice);  // we know slice size fits in 32bit.
```
I'm not sure this is true: it'd be worth an assert at least

jtattermusch replied:
we've allocated read_slice of size 8192 on the previous line. should I still add the assert?

ctiller replied:
oh ffs I can't read

ctiller commented on src/core/support/string.c:
```diff
@@ -173,6 +173,27 @@ int gpr_ltoa(long value, char *string) {
   return i;
 }
 
+int gpr_int64toa(gpr_int64 value, char *string) {
```
I don't see any tests for this function

jtattermusch replied:
Yes, I can add some once I get a thumbs up for adding this PUBLIC function.

jtattermusch commented:
Turns out there were no tests for gpr_ltoa, so I added tests for both gpr_ltoa and gpr_int64toa.
In the process of doing that, it turned out that handling for LONG_MIN and INT64_MIN was broken, so I fixed that. PTAL.

ctiller commented on test/core/support/string_test.c:
```diff
@@ -286,6 +286,53 @@ static void test_strsplit(void) {
   gpr_free(parts);
 }
 
+test_ltoa() {
```
void test_ltoa(void) {

jtattermusch replied:
ah, sorry. made it static void like all the other test methods.

ctiller commented on test/core/support/string_test.c:
```diff
@@ -286,6 +286,53 @@ static void test_strsplit(void) {
+  gpr_free(str);
+}
+
+test_int64toa() {
```
void test_int64toa(void) {

jtattermusch replied:
done

ctiller commented:
Now Linux fails:

```c
src/core/support/string.c: In function 'gpr_ltoa':
src/core/support/string.c:167:12: error: conversion to 'long unsigned int' from 'long int' may change the sign of the result [-Werror=sign-conversion]
     uval = -value;
            ^
src/core/support/string.c:169:12: error: conversion to 'long unsigned int' from 'long int' may change the sign of the result [-Werror=sign-conversion]
     uval = value;
            ^
src/core/support/string.c: In function 'gpr_int64toa':
src/core/support/string.c:193:12: error: conversion to 'gpr_uint64' from 'gpr_int64' may change the sign of the result [-Werror=sign-conversion]
     uval = -value;
            ^
src/core/support/string.c:195:12: error: conversion to 'gpr_uint64' from 'gpr_int64' may change the sign of the result [-Werror=sign-conversion]
     uval = value;
```

jtattermusch commented:
Alright, I did some fix but it looks like this still relies implementation-specific. I will try rewrite the implementation to be 100% compliant with the C spec.

jtattermusch commented:
I got rid of the negation from ltoa and int64toa code, hoping that now everything should be compliant with the spec.

jtattermusch commented:
Weird error building openSSL, I don't see how this could be related.

jtattermusch commented:
Jenkins is sufficiently green.
